File: web/rgbd-js/color_encoder.js
import { NativeByteArray } from './capi_containers.js';

export class NativeColorEncoder {
  constructor(wasmModule, colorCodecType, width, height, targetBitrate, framerate) {
    this.wasmModule = wasmModule;
    this.ptr = this.wasmModule.ccall('rgbd_color_encoder_ctor',
                                     'number',
                                     ['number', 'number', 'number', 'number', 'number'],
                                     [colorCodecType, width, height, targetBitrate, framerate]);
  }

  close() {
    this.wasmModule.ccall('rgbd_color_encoder_dtor', null, ['number'], [this.ptr]);
  }

  encode(yChannel, uChannel, vChannel, keyframe) {
    const yChannelPtr = this.wasmModule._malloc(yChannel.byteLength);
    const uChannelPtr = this.wasmModule._malloc(uChannel.byteLength);
    const vChannelPtr = this.wasmModule._malloc(vChannel.byteLength);
    this.wasmModule.HEAPU8.set(yChannel, yChannelPtr);
    this.wasmModule.HEAPU8.set(uChannel, uChannelPtr);
    this.wasmModule.HEAPU8.set(vChannel, vChannelPtr);
    const byteArrayPtr = this.wasmModule.ccall('rgbd_color_encoder_encode',
                                               'number',
                                               ['number', 'number', 'number', 'number', 'boolean'],
                                               [this.ptr, yChannelPtr, uChannelPtr, vChannelPtr, keyframe]);
    this.wasmModule._free(yChannelPtr);
    this.wasmModule._free(uChannelPtr);
    this.wasmModule._free(vChannelPtr);

    const byteArray = new NativeByteArray(this.wasmModule, byteArrayPtr);
    const bytes = byteArray.toArray();
    byteArray.close();
    return bytes;
  }
}

File: python/pyrgbd/color_encoder.py
from ._librgbd_ffi import lib
from .yuv_frame import YuvFrame
from .capi_containers import NativeByteArray
from .utils import cast_np_array_to_pointer
import numpy as np


class NativeColorEncoder:
    def __init__(self, color_codec_type, width: int, height: int, framerate: int):
        # Setting lib.VP8 assuming since it is the only codec for now.
        # Fix this later when a codec gets added.
        self.ptr = lib.rgbd_color_encoder_ctor(
            color_codec_type, width, height, framerate
        )

    def close(self):
        lib.rgbd_color_encoder_dtor(self.ptr)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def encode(self, yuv_frame: YuvFrame, keyframe) -> np.ndarray:
        with yuv_frame.to_native() as native_yuv_frame:
            return NativeByteArray(
                lib.rgbd_color_encoder_encode(self.ptr, native_yuv_frame.ptr, keyframe)
            ).to_np_array()
